### DISCUSSION TITLE
Decoupling app/logic loop from rendering

### DISCUSSION THREAD

**Discussion by dafteran4** [bevyengine/bevy#6648] [Q&A] [answered]
Hi, I'm new Bevy. I am trying to control my loop via ScheduleRunnerPlugin by setting RunMode::Loop. Which works fine. But as soon as I load RenderPlugin, this doesn't work anymore when RunMode wait is lower than vsync time. This suggests that both rendering and logic are updated in the same loop and app update is waiting for rendering to be ready for the next frame. Is there any way to decouple app loop (input, physics, logic) from rendering? It seems to me these two should be mostly independent processes - rendering running on its own pace and only interacting with app loop as necessary when ready to pick up the last state, but otherwise not blocking the app updates. Any ideas on how to go about this appreciated. Thanks.

**Comment by alice-i-cecile** [comment, marked as answer]
This is "pipelined rendering" :) It's a complex matter, and something that the Bevy dev team is actively pursuing. Not exactly feasible as an end user, but fingers crossed we'll have this enabled by default in the next release.

Follow the work here: https://github.com/bevyengine/bevy/pull/6503

**Reply by dafteran4** [reply]
Thanks for the reference. I am reading through it. From what I gather - and I just might be wrong about this - this still only allows a single app-update per render update. So, even though both run in parallel,  if app/logic finishes early it still has to wait to sync with rendering. Ideally, in scenarios where app/logic updates are significantly faster than rendering, one would want to update the world multiple times for each rendered frame.

**Reply by alice-i-cecile** [reply]
For that, you'll want to use "fixed time steps": https://bevy-cheatbook.github.io/features/fixed-timestep.html

**Reply by dafteran4** [reply]
Thanks again for helping. I experimented with FixedTimestep. Similarly as above, it works as expected as long as I omit the RenderingPlugin. With rendering enabled, however, all fixed timestep events get clustered at the same time point after frame rendering finishes. This unfortunately defeats the purpose of picking up input events and iterating the world with higher frequency than render updates.

**Reply by alice-i-cecile** [reply]
Yep, the model is not ideal there: the work being done in https://github.com/bevyengine/bevy/pull/6587 (and the associated Stageless RFC) will enable us to move towards the model that you're looking for.

**Reply by dafteran4** [reply]
I managed to hack in the desired behavior by changing #6503 `update_rendering()` receiver function to `try_recv()` and skip rendering if previous rendering is not finished yet, thus continuing iterating app updates in the meantime. This somewhat works, but there seems to be some race condition, because sporadically there would be a full frame delay in `CoreStage::PostUpdate`...

**Reply by dafteran4** [reply]
Here is my forked version of #6503 which enables multiple app updates while rendering a single frame. I fixed the sporadic delays, but it probably breaks things in unexpected places where there would be an implicit assumption that one app frame equals one render frame. I'll concede that benefits are marginal: slightly better input latency as events can be received and iterated upon sooner.

[mutli-app-step-while-rendering](https://github.com/dafteran4/bevy/tree/multi-app-step-while-rendering)

**Reply by balintbarna** [reply]
Hi @dafteran4 
Is there an easy way to apply this "patch" to the current version of Bevy so that the app world loop can go faster than rendering?

**Reply by alice-i-cecile** [reply]
You probably want `FixedUpdate`.

**Reply by balintbarna** [reply]
@alice-i-cecile 
Hi Alice, thank you for your reply.
I don't think so, for the same reasons dafteran4 described above. Here's a short description of what I want to do.
I want to create a simulator with an external controller running in the background. The controller needs to receive simulated sensor messages and it sends simulated control messages at a much higher rate than I can run the rendering thread.
For example, the rendering might take 50 ms (20 FPS) while I need the control loop to run somewhere between 125-500 Hz (2-8ms). FixedUpdate seems useless here as it would still only run every 50 ms but then quickly run 25 times. I need the app logic and physics simulation to actually run real-time with sub 10 ms loop times otherwise the controller won't work. The rendering can run even at 10 FPS and it would still be potentially usable.

**Reply by alice-i-cecile** [reply]
Ah, I see. Can you write your own custom schedule using the same sorts of patterns as `FixedUpdate` then? `World::run_schedule` is exposed / designed for this sort of pattern.

**Reply by balintbarna** [reply]
Sorry, I'm not that experienced with bevy yet but my understanding is that I would need to put world.run_schedule(MySchedule) inside a system. like:
```
fn main() {
  ...
  app.add_systems(Update, run_my_schedule)
}

fn run_my_schedule(world: &mut World) {
  for _ in 0..10 {
    world.run_schedule(MySchedule);
  }
}
```
or something like this'

But then basically I would just run MySchedule 10 times and then it would sync with the rendering pipeline and still block.

Maybe if there's a way to keep my schedule running until the renderer finishes? Is there an example for how to do that? Maybe like a flag saying if the current frame rendering fininshed?

**Reply by alice-i-cecile** [reply]
Can you use a channel or other method of communication to wait and hear if the rendering is complete?

**Reply by balintbarna** [reply]
I'm not sure what you mean, the rendering is handled by the pipelined rendering plugin from the PR mentioned earlier in this thread, right? I don't know how to get the information from there. With dafteran4's hack I wouldn't need to because the main schedule would keep looping and not block waiting for the render to complete and I guess your suggestion could have the same effect by running a specific schedule instead of the entire main schedule repeatedly in a loop until the last frame is rendered and then it could finish the schedule -> do the extract -> start the new frames, but in my understanding for that I would need to know when the last frame finished rendering, otherwise I wouldn't know how long to keep looping on my schedule, does that make sense?

**Reply by alice-i-cecile** [reply]
Yeah, in order to get the information from there, I would try using https://docs.rs/crossbeam/latest/crossbeam/channel/index.html

**Reply by balintbarna** [reply]
I appreciate your help. I can't get it to work and I don't know why. I used the built-in channels instead of crossbeam but I imagine that shouldn't be a problem. Here's the plugin which I wrote to use for the looped parallel execution:
```
use std::sync::{
    mpsc::{channel, Receiver, Sender},
    Arc, Mutex,
};

use bevy::{
    app::{App, Plugin, PostUpdate},
    ecs::schedule::{ExecutorKind, ScheduleLabel},
    prelude::{IntoSystemConfigs, Res, Resource, Schedule},
    render::{Render, RenderApp, RenderSet},
};

#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
pub struct RealTimeControl;

pub struct RealTimeSimulation;

impl Plugin for RealTimeSimulation {
    fn build(&self, app: &mut App) {
        let mut sched: Schedule = Schedule::new(RealTimeControl);
        sched.set_executor_kind(ExecutorKind::MultiThreaded);
        let (tx, rx) = channel();
        app.add_schedule(sched)
            .insert_resource(RenderFinishedReceiver(Arc::new(Mutex::new(rx))))
            .add_systems(PostUpdate, run_physics_while_rendering);
        app.sub_app_mut(RenderApp)
            .insert_resource(RenderFinishedSender(tx));
        // .add_systems(Render, signal_render_finished.in_set(RenderSet::Cleanup));
    }
}

#[derive(Resource)]
struct RenderFinishedReceiver(Arc<Mutex<Receiver<()>>>);

fn run_physics_while_rendering(finished: Res<RenderFinishedReceiver>) {
    if finished.0.lock().unwrap().iter().last().is_some() {
        println!("Rendering finished again!");
    }
    // while world.resource::<Events<RenderFinishedEvent>>().len() == 0 {
    //     world.run_schedule(MultiMain);
    // }
    // world.resource_mut::<Events<RenderFinishedEvent>>().clear();
    // print!("Finished run phy-------------------------------------");
}

#[derive(Resource)]
struct RenderFinishedSender(Sender<()>);

fn signal_render_finished(finished: Res<RenderFinishedSender>) {
    if finished.0.send(()).is_err() {
        println!("Could not send finished signal!")
    }
    println!("Rendering finished-------------------------------------");
}
```
If I remove the line that inserts the sender as a resource the rendering starts up normally, however with that resource inserted I just get a black screen in the app window and it's frozen (get  force quit popup).

If I understand correctly your suggestion was to create a channel which let's me send a signal from the rendering subapp. Basically I created a channel, inserted the receiver to the main app where an update system can check the RX to see if a frame has been rendered. I also inserted the TX to the render subapp where a system on the cleanup set would send a signal. I have no idea if this is how you meant it but it seems plausible to me, but for some reason breaks the app.

**Reply by alice-i-cecile** [reply]
Yeah, that's the basic idea that I had in mind. I'll check if anyone on Discord has any better ideas.

**Reply by balintbarna** [reply]
I think it’s a good idea but I don’t know why it breaks the program to add a resource to a sub app. Is that normal?

**Reply by alice-i-cecile** [reply]
No, that's quite unusual. Not sure what's going on there TBH. There's a few more suggestions from [Discord](https://discord.com/channels/691052431525675048/692648082499829760/1302998807331737612): a custom executor, running the simulation in an async task, making sure your rendering is fully on a separate thread.

**Reply by thebluefish** [reply]
I would wager that you're blocking the thread by trying to `iter` an empty channel. `iter` and `recv` both **block** the thread until a message is received, you probably want `try_recv` instead.

**Reply by balintbarna** [reply]
Right! Maybe try_iter 
I’ll try it later today
I guess it doesn’t cause an issue if I don’t add the resource because the channel just gets closed

**Reply by balintbarna** [reply]
It's working:

```
use std::sync::{
    mpsc::{channel, Receiver, Sender},
    Arc, Mutex,
};

use bevy::{
    app::{App, Plugin, PostUpdate},
    ecs::schedule::{ExecutorKind, ScheduleLabel},
    prelude::{IntoSystemConfigs, Res, Resource, Schedule, World},
    render::{Render, RenderApp, RenderSet},
};

#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
pub struct RealTimeControl;

pub struct RealTimeSimulation;

impl Plugin for RealTimeSimulation {
    fn build(&self, app: &mut App) {
        let mut sched: Schedule = Schedule::new(RealTimeControl);
        sched.set_executor_kind(ExecutorKind::MultiThreaded);
        let (tx, rx) = channel();
        tx.send(()).unwrap();
        app.add_schedule(sched)
            .insert_resource(RenderFinishedReceiver(Arc::new(Mutex::new(rx))))
            .add_systems(PostUpdate, run_physics_while_rendering);
        app.sub_app_mut(RenderApp)
            .insert_resource(RenderFinishedSender(tx))
            .add_systems(Render, signal_render_finished.in_set(RenderSet::Cleanup));
    }
}

#[derive(Resource)]
struct RenderFinishedReceiver(Arc<Mutex<Receiver<()>>>);

fn run_physics_while_rendering(world: &mut World) {
    while world
        .resource::<RenderFinishedReceiver>()
        .0
        .lock()
        .unwrap()
        .try_iter()
        .last()
        .is_none()
    {
        world.run_schedule(RealTimeControl);
    }
}

#[derive(Resource)]
struct RenderFinishedSender(Sender<()>);

fn signal_render_finished(finished: Res<RenderFinishedSender>) {
    if finished.0.send(()).is_err() {
        log::error!("Could not send signal after render finished!")
    }
}
```

Thank you all for your help!

Now the problem is only that Avian - the physics engine running on the new schedule - seems to be moving extremely fast. The cube falls down _instantly_ instead of the regular pace. I will try to figure out why that is :)
